File: masar_miraaya/masar_miraaya/report/sales_forecast/sales_forecast.js
// Copyright (c) 2025, KCSC and contributors
// For license information, please see license.txt

frappe.query_reports["Sales Forecast"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": "Company",
			"fieldtype": "Link",
			"options": "Company",
		},
		{
			"fieldname": "from_date",
			"label": "From Date",
			"fieldtype": "Date",
			"default": "Today - 90 days",
			"reqd": 1
		},
		{
			"fieldname": "to_date",
			"label": "To Date",
			"fieldtype": "Date",
			"default": "Today",
			"reqd": 1
		},
		{
			"fieldname": "based_on_document",
			"label": "Based On Document",
			"fieldtype": "Select",
			"options": "Sales Order\nSales Invoice",
			"default": "Sales Invoice",
			"reqd": 1
		},
		{
			"fieldname": "item_code",
			"label": "Item Code",
			"fieldtype": "Link",
			"options": "Item",
		},
		{
			"fieldname": "item_group",
			"label": "Item Group",
			"fieldtype": "Link",
			"options": "Item Group"
		},
		{
			"fieldname": "brand",
			"label": "Brand",
			"fieldtype": "Link",
			"options": "Brand"
		},
		{
			"fieldname": "months_for_forecast",
			"label": "Months for Forecast",
			"fieldtype": "Int",
			"default": 3
		},
		{
			"fieldname": "periodicity",
			"label": "Periodicity",
			"fieldtype": "Select",
			"options": "Monthly\nQuarterly\nYearly",
			"default": "Monthly",
			"reqd": 1
		},
		{
			"fieldname": "forecast_method",
			"label": "Forecast Method",
			"fieldtype": "Select",
			"options": "Simple Average\nWeighted Average\nLinear Regression",
			"default": "Simple Average"
		}
	]
};

File: masar_miraaya/masar_miraaya/report/sales_forecast/sales_forecast.py
# Copyright (c) 2025, KCSC and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from datetime import datetime


def execute(filters=None):
	data = get_data(filters)
	columns = get_columns(data)
	return data, columns

def get_data(filters):
	if filters.get("based_on_document"):
		from_doc = filters.get("based_on_document")
		from_child_doc = from_doc + " Item"
	if filters.get("periodicity"):
		periodicity = filters.get("periodicity")
		if periodicity == "Monthly":
			group_expr = "DATE_FORMAT(tsi.posting_date, '%%Y-%%m')"
		elif periodicity == "Quarterly":
			group_expr = "CONCAT(YEAR(tsi.posting_date), '-Q', QUARTER(tsi.posting_date))"
		elif periodicity == "Yearly":
			group_expr = "YEAR(tsi.posting_date)"
	sql = frappe.db.sql(f"""
		SELECT 
			tsii.item_code,
			ti.item_name,
			ti.item_group,
			ti.brand,
			{group_expr} AS period,
			SUM(tsii.qty) AS total_qty
		FROM `tab{from_doc}` tsi
		INNER JOIN `tab{from_child_doc}` tsii ON tsi.name = tsii.parent
		INNER JOIN `tabItem` ti ON tsii.item_code = ti.name
		GROUP BY tsii.item_code
		ORDER BY tsii.item_code
	""", as_dict=True)

	return sql

def get_columns(data):
	base_columns = [
		{"label": _("Item Code"), "fieldname": "item_code", "fieldtype": "Link", "options": "Item", "width": 120},
		{"label": _("Item Name"), "fieldname": "item_name", "fieldtype": "Data", "width": 200},
		{"label": _("Item Group"), "fieldname": "item_group", "fieldtype": "Link", "options": "Item Group", "width": 150},
		{"label": _("Brand"), "fieldname": "brand", "fieldtype": "Link", "options": "Brand", "width": 150},
	]
	
	period_keys = set()
	for row in data:
		for key in row:
			if key not in {"item_code", "item_name", "item_group", "brand"}:
				period_keys.add(key)

	# Sort period keys (e.g. '2024-01', '2024-02', etc.)
	period_columns = [{"label": p, "fieldname": p, "fieldtype": "Float", "width": 120} for p in sorted(period_keys)]
	return base_columns + period_columns